### ISSUE TITLE
Fail to use navigation property of intermediate result

### ISSUE THREAD

**Issue by shameleo** [linq2db/linq2db#4274]
Linq2db honors association attribute when class is used as table representation, but not as intermediate result.

```cs
using LinqToDB;
using LinqToDB.Data;
using LinqToDB.Mapping;
using System.IO;
using System.Linq;
using System.Linq.Expressions;

namespace l2db_bug
{
    class Db : DataConnection
    {
        public Db(string path) :
            base("SQLite", $@"Data Source = {path}; foreign keys = true; Version = 3;")
        { }
    }

    public class Service
    {
        public int Id { get; set; }
        public int? IdClient { get; set; }

        //[Association(ExpressionPredicate = nameof(ClientPredicate))]
        [Association(ThisKey = nameof(IdClient), OtherKey = nameof(l2db_bug.Client.Id))]
        public Client Client { get; set; }

        //static public Expression<Func<Service, Client, bool>> ClientPredicate => (s, c) => s.IdClient == c.Id;
    }

    public class Client
    {
        public int Id { get; set; }
    }

    class Program
    {
        static void Main(string[] args)
        {
            string dbPath = "db.db";

            using (File.Create(dbPath)) { }

            using (var db = new Db(dbPath))
            {
                db.CreateTable<Service>();
                db.CreateTable<Client>();

                //// works
                // var q_serv = (
                //     from serv in db.GetTable<Service>()
                //     select serv
                // );

                //// must "connect" navigation property to property of source, otherwise will get exception later
                // var q_serv = (
                //     from serv in db.GetTable<Service>()
                //     select new Service
                //     {
                //         Id = serv.Id,
                //         IdClient = serv.IdClient,
                //         Client = serv.Client         // <== here
                //     }
                // );

                //// NO WAY to do it here. And association attribute is ignored now
                var q_serv = (
                    from serv in db.GetTable<Service>()
                    group serv by new { serv.Id, serv.IdClient } into gr
                    select new Service      // reusing Service type for simplicity
                    {
                        Id = gr.Key.Id,
                        IdClient = gr.Key.IdClient
                    }
                );

                var q_result = (
                    from serv in q_serv
                    where serv.Client.Id == 1
                    select serv
                );

                // 'LinqToDB.LinqToDBException' occurred in linq2db.dll: 
                // 'Member 'Client' not found in type 'Service'.'
                var result = q_result.ToList();
            }
        }
    }
}
```
<details>
<summary>full exception stack trace</summary>

```
Exception has occurred: CLR/LinqToDB.LinqToDBException
An unhandled exception of type 'LinqToDB.LinqToDBException' occurred in linq2db.dll: 'Member 'Client' not found in type 'Service'.'
   at LinqToDB.Linq.Builder.SelectContext.GetProjectedExpression(MemberInfo memberInfo, Boolean throwOnError)
   at LinqToDB.Linq.Builder.SelectContext.ProcessMemberAccess[T,TContext](TContext context, Expression expression, MemberExpression levelExpression, Int32 level, Func`7 action)
   at LinqToDB.Linq.Builder.SelectContext.ConvertToIndexInternal(Expression expression, Int32 level, ConvertFlags flags)
   at LinqToDB.Linq.Builder.SelectContext.ConvertToIndex(Expression expression, Int32 level, ConvertFlags flags)
   at LinqToDB.Linq.Builder.SubQueryContext.ConvertToSql(Expression expression, Int32 level, ConvertFlags flags)
   at LinqToDB.Linq.Builder.ExpressionContext.ConvertToSql(Expression expression, Int32 level, ConvertFlags flags)
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertToSql(IBuildContext context, Expression expression, Boolean unwrap, ColumnDescriptor columnDescriptor, Boolean isPureExpression)
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertCompare(IBuildContext context, ExpressionType nodeType, Expression left, Expression right)
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertPredicate(IBuildContext context, Expression expression)
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSearchCondition(IBuildContext context, Expression expression, List`1 conditions)
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildWhere(IBuildContext parent, IBuildContext sequence, LambdaExpression condition, Boolean checkForSubQuery, Boolean enforceHaving)
   at LinqToDB.Linq.Builder.WhereBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.MethodCallBuilder.BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.ExpressionBuilder.Build[T]()
   at LinqToDB.Linq.Query`1.CreateQuery(ExpressionTreeOptimizationContext optimizationContext, ParametersContext parametersContext, IDataContext dataContext, Expression expr)
   at LinqToDB.Linq.Query`1.GetQuery(IDataContext dataContext, Expression& expr, Boolean& dependsOnParameters)
   at LinqToDB.Linq.ExpressionQuery`1.GetQuery(Expression& expression, Boolean cache, Boolean& dependsOnParameters)
   at LinqToDB.Linq.ExpressionQuery`1.System.Collections.Generic.IEnumerable<T>.GetEnumerator()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at l2db_bug.Program.Main(String[] args) in /run/media/shameleo/bbd007ef-fd51-494e-a48e-276e82e74075/l2db_navigation_issue/Program.cs:line 89
```
</details>

And no, explicit join is bad workaround. Navigation properties make it much easier to create composable methods to build queries with. If it's not available, then I should replace (pseudocode)

```
Service GetServiceGrouped(db)
{
    var q_serv = ...;
    ...
    return q_serv;
}
```
with something like
```
ServiceWithClient GetServiceGrouped(db, shouldJoinClient)
{
    var q_serv = ...;
    var q_joined = shouldJoinClient 
        ? (
            from serv in q_serv
            join cl in db.Clients on serv.IdClient equals cl.Id
            select new ServiceWithClient
            {
                Service = serv,        // don't you ever use Service.Client later!
                Client = cl
            }
        )
        : (
            from serv in q_serv
            select new ServiceWithClient
            {
                Service = serv, 
                Client = null
            }
        );
        
    ...
    return q_joined;
}
```
which is pretty ugly and grows uglier if need more joins and need to maintain all this `shouldJoinClient` and `ServiceWithClient` stuff through several levels. If you would want to get rid of at least `shouldJoinClient` and replace it with unconditional left join, than you should know it's not always possible. For instance, I have cross database joins, so I can't tell to DBMS that Client will be unique for given Service, as a result it won't be able to remove join from execution plan if it is not actually used.

---
So, conclusion: being unable to use navigation properties whenever nessesary is pretty sad. Looking forward for fix!
PS: is it related? https://github.com/linq2db/linq2db/issues/2461

**Comment by MaceWindu** [comment]
We will check if it is possible to fix with current parser, otherwise it will be fixed we new linq parser we are working on now

**Comment by MaceWindu** [comment]
Fixed in `6.0.0-preview.1`
